java/vector/src/main/java/org/apache/arrow/vector/complex/NonNullableStructVector.java

```diff
@@ -303,7 +303,7 @@ public boolean equals(int index, ValueVector to, int toIndex) {
     if (to == null) {
       return false;
     }
-    if (this.getClass() != to.getClass()) {
+    if (!this.getField().getType().equals(to.getField().getType())) {
```

Comment:
we are not comparing the names of the children. for a struct, that seems incorrect - but the overhead of doing so will be high.

Reply:
Do we need worry about children names? I think two union vectors will equal if they have same type vectors and elements?

Reply:
for eg. I think: 

struct s1 { int a, long b} != struct s2 { int b, long a }

i.e the names of the children are part of the type. I'll verify what the cpp code does

Reply:
The cpp code compares the name of the children too.

compare.cc has 
ArrayEqualsImpl ->BaseDataEquals -> TypeEquals
TypeEqualsVisitor::visitChildren() -> Field::Equals -> compares name

Reply:
Thanks for your confirm. fixed now and also added a ut.